### PR TITLE
Implement the Tarok game

### PR DESCRIPTION
**Pull request by inejc**
This PR adds the Slovenian Tarok game as implemented in https://github.com/semanticweights/tarok (a full game without announcements).

Todo:
- [x] port the main files from https://github.com/semanticweights/tarok/tree/master/tarok/src
- [x] implement custom state de/serialization (game is implicitly stochastic)
- [x] port custom python bindings from https://github.com/semanticweights/tarok/blob/master/tarok/src/py_bindings.cpp
- [x] port tests from https://github.com/semanticweights/tarok/tree/master/tarok/test
- [x] fix bonus for valat game
- [x] write readme (link the original repo)
- [x] update python integration tests (playthrough test)
- [x] add the game to the game wiki list

### CODE REVIEW

**Comment by inejc** [comment]
@lanctot this PR is now deemed finished. The environment is tested quite extensively, although there are some minimal missing tests for rewards which I will add in a separate PR. In addition to that, the implementation of announcements (https://www.pagat.com/tarot/sltarok.html#announcements) is also missing and will come in a future PR (the game is fully playable nevertheless as announcements are an optional addition).

**Comment by lanctot** [comment]
Very nice! Quite happy to see this ready! Thanks for all the continued work on it.

We have a few large-ish PRs in the stack that we've been had some trouble finding time for, so it might take some time, but we'll keep you posted. I hope once we clear all the small ones that we can each take a larger one get it done within the next few weeks.

**Comment by inejc** [comment]
Thanks for letting me know, I might add the missing tests to this PR then.

**Comment by inejc** [comment]
I ran the [open_spiel/python/examples/benchmark_games.py](https://github.com/deepmind/open_spiel/blob/8cfdc9d1cab6709199a62d417732b076ef69cb13/open_spiel/python/examples/benchmark_games.py) script to sanity check the performance of the implementation (not sure how comparable the below games are but at least this serves as a check that we aren't magnitudes slower). Tarok is by default played with three players.

```
$ python open_spiel/python/examples/benchmark_games.py
---
Results for following benchmark configuration:
time_limit = 10.0
give_up_after = 100
---
                          Game  msec/rollout   msec/move  Give ups/rollouts  Time elapsed [sec]
3                       bridge    194.344768    3.056845           0.000000           10.105928
21                      hearts      0.372413    0.003687           1.000000           10.000042
46                        skat      0.218475    0.003550           0.000000           10.000035
48                       tarok      0.057921    0.004729           0.000000           10.000001
50              tiny_bridge_2p      0.021224    0.004276           0.000000           10.000003
51              tiny_bridge_4p      0.057631    0.004544           0.000000           10.000070
```

**Review by flostim — COMMENTED** [review]
Hey, this looks great and extremely well tested. Thanks for all the work!

Can I ask why do you prefer implicit stochasticity to the way that Skat & Bridge are implemented (i.e. each card dealt is an action by the chance player with deterministic outcome)?

File: open_spiel/python/pybind11/pyspiel.cc [hunk 1]
```diff
@@ -566,6 +567,9 @@ PYBIND11_MODULE(pyspiel, m) {
   init_pyspiel_games_bridge(m);  // Game-specific functions for bridge.
   init_pyspiel_observer(m);      // Observers and observations.
 
+  // List of custom game bindings.
+  init_pyspiel_tarok(m);
+
```
[the remark this answers is not in the document]
Comment (inejc) [reply open_spiel/python/pybind11/pyspiel.cc]: I moved the binding files and generally used the same approach as for other games.

File: open_spiel/games/tarok_test.cc [hunk 2]
```diff
@@ -0,0 +1,1445 @@
...
+  // shuffles should be the same when recreating a game with the same seed
+  auto state3 = game->NewInitialTarokState();
+  state3->ApplyAction(0);
+  auto state4 = game->NewInitialTarokState();
+  state4->ApplyAction(0);
+  SPIEL_CHECK_EQ(state1->PlayerCards(0), state3->PlayerCards(0));
+  SPIEL_CHECK_EQ(state2->PlayerCards(0), state4->PlayerCards(0));
+}
+
+void DealtCardsSizeTest() {
```
[the remark this answers is not in the document]
Comment (inejc) [reply open_spiel/games/tarok_test.cc]: I added some more tests.

**Comment by inejc** [comment]
Hey! The reasoning was that implicit implementation is easier (mostly meaning that less code had to be written + it's more readable). In addition, I don't think that any algorithm that relies on the explicit game tree would be viable for a game this large. One drawback is that we needed to add the internal RNG and implement custom serialization so not sure if we ended up writing less code after all :). Hope the answer helps, please let me know what you think.

**Comment by flostim** [comment]
> Hey! The reasoning was that implicit implementation is easier (mostly meaning that less code had to be written + it's more readable). In addition, I don't think that any algorithm that relies on the explicit game tree would be viable for a game this large. One drawback is that we needed to add the internal RNG and implement custom serialization so not sure if we ended up writing less code after all :). Hope the answer helps, please let me know what you think.

Sounds reasonable! Could you add a quick note about that in the readme. It's already well documented in the code. Are you okay with us changing this in the future?

**Comment by inejc** [comment]
> Could you add a quick note about that in the readme.

Done.

> Are you okay with us changing this in the future?

Do you mean changing to explicit stochasticity? Sure, I don't mind that. Would be interested in what the pros for the change would be to better understand the framework internals.

**Comment by flostim** [comment]
> > Could you add a quick note about that in the readme.
> 
> Done.
> 
> > Are you okay with us changing this in the future?
> 
> Do you mean changing to explicit stochasticity? Sure, I don't mind that. Would be interested in what the pros for the change would be to better understand the framework internals.

I think mainly because the other card games are implemented that way and it allows to play around more with the game tree. E.g. you could run games where some of the dealings are fixed e.g. player 1 always get card x,y and z.

But mainly thanks again for all the work you put into this! I'll get around importing it.

**Comment by inejc** [comment]
> I think mainly because the other card games are implemented that way and it allows to play around more with the game tree. E.g. you could run games where some of the dealings are fixed e.g. player 1 always get card x,y and z.

That's a useful case that I haven't thought about! I any case, I will keep working on the environment so might end up changing it myself.

Thanks for the review, really happy to see the PR imported!

**Pull request merged**